File: leet_1046_last_stone___priorityQ.ts
function lastStoneWeight(stones: number[]): number {

    // impl. priority queue

    let y: number, x: number
    while (stones.length > 1) {
        stones.sort((a, b) => b - a)
        // console.log(stones)
        y = stones.shift()
        x = stones.shift()
        stones.push(y - x)
    }
    return stones.length ? stones[0] : 0
};

File: leet_1046_last_stone___priorityQ.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        while len(stones) > 1:
            stones.sort()
            y = stones.pop()
            x = stones.pop()
            stones.append(y - x)
        return 0 if len(stones) == 0 else stones[0]
